README.md
# Littlechef-Rackspace

[![Build Status](https://travis-ci.org/tildedave/littlechef-rackspace.png)](https://travis-ci.org/tildedave/littlechef-rackspace)

Deploy chef to Rackspace Cloud Servers powered by OpenStack without a Chef Server.

Replaces `knife rackspace` for users of LittleChef.

Built on top of the excellent [libcloud](http://libcloud.org/) and [littlechef](https://github.com/tobami/littlechef) libraries.

## Installation Instructions

You must have libyaml-dev installed (C library required by PyYAML).

```
pip install littlechef-rackspace
```

## Configuration

`littlechef-rackspace` supports a number of commands for managing your Cloud Resources.

* create
* rebuild
* list-servers
* list-images
* list-networks
* list-flavors

To do any of these, commands you must specify `username`, `key`, and `region`.

You can specify arguments on the command-line or in a `rackspace.yaml` file.x
Your `rackspace.yaml` might look something like:

```yaml
secrets-file: secrets.cfg
image: 5cebb13a-f783-4f8c-8058-c4182c724ccd
flavor: 2
public_key: bootstrap.pub
private_key: bootstrap
region: dfw
environment: preprod
```

* `region`: You can choose any Rackspace Cloud region (DFW, ORD, IAD,
  SYD, LON, HKG).  For LON servers you must have an enabled account (sign up
  at rackspace.co.uk).

### Putting Your Secrets in a Secrets File

You can specify a `secrets-file` in the `rackspace.yaml` that contains your username and api key.
This allows your source repository to track common configuration (images, flavors) but not secrets.
The `secrets.cfg` file is a Python configuration file and secrets will be read from under the
`[DEFAULT]` configuration section. Here's an example `secrets.cfg` file:

```
[DEFAULT]
username=myrackspaceusername
key=myrackspaceapikey
```

## Rackspace Create

```
fix-rackspace create \
    --username <username> \
    --key <api_key> \
    --region <region, must be DFW or ORD> \
    --image 5cebb13a-f783-4f8c-8058-c4182c724ccd \
    --flavor 2 \
    --name "test-creation" \
    --public-key <public_key_file> \
    --private-key <private_key_file> \
    --runlist "role[web],recipe[security-updates]" \
    --plugins "save_network,save_cloud" \
    --post-plugins "mark_node_as_ready"
```

### Arguments

* `image`: Image to use (e.g. Ubuntu 12.04)
* `flavor`: Flavor to use (e.g. 2GB, 4GB, 8GB)
* `name`: Name of the newly created node in the Rackspace Cloud Control panel.  Also sets the initial
  hostname.  You will need to set up DNS yourself.
* `public_key`: Public key used to authorize root SSH access
* `private_key`: Private key used to authenticate as root to the newly created node
* `environment`: Sets the `chef_environment` on a newly created node
* `runlist`: Comma separated list of the Chef runlist to execute on the node.
* `plugins`: Comma separated list of littlechef plugins.  Plugins are executed after chef deploy but
  before your recipe run.  If your recipes depend on the `cloud` attribute being set, you can specify a custom
  littlechef `save_cloud` plugin that uses ohai to save data to the node before your cookbooks are run.
* `post-plugins`: Comma separated list of littlechef plugins.  These plugins are executed after the initial
  chef run.  They can be used to mark that a node is ready to go into rotation (for example).
* `skip-opscode-chef`: Don't run `deploy_chef` to install chef with opscode packages (only on command line)
* `use-opscode-chef`: '0' or '1' based on whether to run `deploy_chef` after initial node startup (defaults to 1).
  Useful when you are installing your own chef packages through a plugin.

## Rackspace Rebuild

```
fix-rackspace rebuild \
    --username <username> \
    --key <api_key> \
    --region <region, must be DFW or ORD> \
    --image 5cebb13a-f783-4f8c-8058-c4182c724ccd \
    --name "test-rebuild" \
    --public-key <public_key_file> \
    --private-key <private_key_file> \
    --runlist "role[web],recipe[security-updates]" \
    --plugins "save_network,save_cloud" \
    --post-plugins "mark_node_as_ready"
```

### Arguments

Arguments are as for create, with the exception of:

* `name`: Name of the node to rebuild.

## Rackspace List Servers

List the servers with your associated region.  Useful for determining which servers you want to rebuild.

```
4e7d282f-3827-4c71-bb1d-7519e0543c4b     server1                  5.6.7.8
30cd9fef-4302-47fa-9f05-3fc8cc35a0b7     server2                  12.13.14.15
```

## Rackspace List Images

List the images with your associated region.  Handy for finding the image you want to create.

```
$ fix-rackspace list-images
bca91446-e60e-42e7-9e39-0582e7e20fb9       Fedora 16 (Verne)
17282573-81b3-4a2d-80e4-7fcf16d4ac08       Windows Server 2012 (with updates) + SQL Server 2012 Web SP1
a620e9ee-cd4c-47a0-b909-042cb5058488       Windows Server 2012 (with updates) + SQL Server 2012 Standard SP1
3071dff8-6564-481e-a9c0-2c67f0b8cbf0       Windows Server 2012 (with updates)
# etc
```

## Rackspace List Flavors

```
$ fix-rackspace list-flavors
2         512MB Standard Instance
3         1GB Standard Instance
4         2GB Standard Instance
5         4GB Standard Instance
6         8GB Standard Instance
7         15GB Standard Instance
8         30GB Standard Instance
```

## Reducing Command-Line Boilerplate With Templates

In practice many arguments are grouped together for creates.  For example, you may have a staging install in the DFW datacenter, but a production install in the ORD datacenter.  These datacenters all use different private network identifiers.  Additionally, you may have several types of node: web, application server, database, each with different plugins or runlists.

To support this hierarchy in arguments and reduce "what were those arguments anyways?" syndrome, `littlechef-rackspace` supports _templates_.  These are specified in your rackspace.yml.

```yaml
templates:
  base:
    runlist:
      - recipe[test]
      - recipe[test2]
  web:
    runlist:
      - recipe[web]
  preprod:
    region: dfw
    environment: preprod
    networks:
      - 00000000-0000-0000-0000-000000000000
      - 3d443c50-a45e-11e3-a5e2-0800200c9a66
  production:
    region: ord
    environment: production
    networks:
      - 00000000-0000-0000-0000-000000000000
      - 5df00920-a45e-11e3-a5e2-0800200c9a66
```

Once a template is defined you can use it as part of node creation by specifying it after the 'create'.

```
# Applies 'web' and 'preprod' templates to new node 'web-n01.preprod'
fix-rackspace create --name web-n01.preprod web preprod

# Applies 'web' and 'production' templates to new node 'web-n01.prod'
fix-rackspace create --name web-n01.prod web production
```

List-based arguments will be combined if multiple templates apply.  For example,
given the following template:

```yaml
templates:
  base:
    networks:
      - 00000000-0000-0000-0000-000000000000
  preprod:
    region: dfw
    environment: preprod
    networks:
      - 3d443c50-a45e-11e3-a5e2-0800200c9a66
```

The following command will create a node with two networks,
`00000000-0000-0000-0000-000000000000` and `3d443c50-a45e-11e3-a5e2-0800200c9a66`.

```
fix-rackspace create --name base-n01.preprod base preprod
```

### Notes

The server is created with your public key file in the `/root/.ssh/authorized_keys`.
I highly recommended that you disable root password login as part of your chef
recipes!

## Tips for OS X

Mac users will most likely encounter the following libcloud error:

```
No CA Certificates were found in CA_CERTS_PATH
```

This can be remedied by executing the following:

```
$ brew install curl-ca-bundle
$ export SSL_CERT_FILE=/usr/local/opt/curl-ca-bundle/share/ca-bundle.crt
```

More information about this error can be found at [libcloud.apache.org](http://libcloud.apache.org/docs/ssl-certificate-validation.html)


Changelog.md
## 1.0 (pre)

* Add "rebuild" and "list-servers" commands
* Display performance flavors better (Samuel Toriel)
* Support creating IAD and HKG public cloud resources
* Add --dry-run argument for showing what arguments
  a node will be created with, but not actually executing
  the creation.
* Deprecate 'rackspace' section of config.cfg; LittleChef has deprecated
  the use of a config.cfg.  Future configuration for littlechef-rackspace
  should use rackspace.yaml.
* Add template support to creating nodes: templates are groups of arguments
  that are specified on the command line after 'create' and should reduce
  the boilerplate associated with spinning up a new node.
* BREAKING: Remove 'hostname' argument.  Nodes are
  created with the same hostname as their API name.
* BREAKING: littlechef 1.6.1 is required.  Your littlechef configuration
  needs to be specified in littlechef.cfg now (not config.cfg), and you
  must have an environments/ directory.

## 0.6 (2013-04-25)

* Support creating SYD resources (if your account has Sydney OpenStack
  enabled)

## 0.5 (2013-04-05)

* list-networks command to view custom networks on your account
* create command now takes a networks list
* Fixed regression introduced in 0.4 where not specifying opscode
  Chef would not deploy chef on new nodes

## 0.4 (2013-03-27)

* Python 2.6 support
* --skip-opscode-chef and --use-opscode-chef options to allow not using
  Opscode Chef Repositories on newly bootstrapped nodes

## 0.3 (2013-03-24)

* --hostname to create node file with specified hostname
* --plugins and --post-plugins for executing plugins before and after chef run

## 0.2 (2013-03-20)

* --environment to create nodes in a specific Chef Environment
* Fixed some errors in configuration parsing

## 0.1 (2013-02-05)

* Initial release, nothing works


littlechef_rackspace/runner.py
from optparse import OptionParser
from fabric.utils import abort

import ConfigParser
import os
import littlechef
import yaml

from api import RackspaceApi
from deploy import ChefDeployer
from commands import (RackspaceCreate,
                      RackspaceListImages,
                      RackspaceListFlavors,
                      RackspaceListNetworks,
                      RackspaceRebuild,
                      RackspaceListServers)


def get_command_classes():
    return [RackspaceCreate,
            RackspaceRebuild,
            RackspaceListImages,
            RackspaceListFlavors,
            RackspaceListNetworks,
            RackspaceListServers]


class FailureMessages:

    INVALID_REGION = ("Must specify a valid region "
                      "'dfw', 'ord', 'iad', 'syd', 'hkg', 'lon'.")

    NEED_API_KEY = ("Must specify username, API key, and region "
                    "on command line or in [rackspace] configuration "
                    "section of config.cfg")

    MISSING_REQUIRED_ARGUMENTS = "Missing required arguments"

    MUST_SPECIFY_PUBLICNET = ("Must specify PublicNet in networks "
                              "list (id=00000000-0000-0000-0000-000000000000)")


class RackspaceOptionParser(OptionParser):
    def print_help(self, file=None):
        OptionParser.print_help(self, file)
        print("\nAvailable commands:\n")
        for command_class in get_command_classes():
            print("   {0}\t{1}".format(command_class.name,
                                       command_class.description))
        print("")


parser = RackspaceOptionParser()
parser.add_option("-I", "--image", dest="image",
                  help="Image ID")
parser.add_option("-f", "--flavor", dest="flavor",
                  help="Flavor ID")
parser.add_option("-A", "--username", dest="username",
                  help="Rackspace Username")
parser.add_option("-N", "--name", dest="name",
                  help="Node name (will become hostname for node)")
parser.add_option("-K", "--key", dest="key",
                  help="Rackspace API Key")
parser.add_option("-R", "--region", dest="region", default="",
                  help="Region for provisioning (required for OpenStack)")
parser.add_option("-a", "--public-key", dest="public_key",
                  help="Public Key File for Bootstrapping")
parser.add_option("-i", "--private-key", dest="private_key",
                  help="Private Key File for Bootstrapping")
parser.add_option("-r", "--runlist", dest="runlist",
                  help=("Node runlist delimited by commas, e.g. "
                        "'role[web],recipe[db]'"))
parser.add_option("-e", "--env", dest="environment",
                  help="Environment for newly created node",
                  default=None)
parser.add_option("-p", "--plugins", dest="plugins",
                  help=("Plugins to execute after chef bootstrapping but "
                        "before chef run e.g, 'save_cloud,save_hosts'"),
                  default=None)
parser.add_option("-P", "--post-plugins", dest="post-plugins",
                  help=("Plugins to execute after chef run. "
                        "e.g, 'mark_node_as_provisioned'"),
                  default=None)
parser.add_option("--skip-opscode-chef", action="store_false",
                  dest="use-opscode-chef")
parser.add_option("--dry-run", action="store_true", dest="dry_run",
                  help="When creating a node, do not actually create/deploy")
parser.add_option("--use-opscode-chef", type="int", dest="use-opscode-chef",
                  help=("Integer argument with whether or not to use "
                        "the OpsCode Chef repositorities "
                        "(installed with 'fix deploy_chef')"),
                  default=None)
parser.add_option("-n", "--networks", dest="networks",
                  help="Comma separated list of network ids to \
                          create node with (PublicNet is required)",
                  default=None)


class Runner(object):
    def _read_littlechef_config(self):
        try:
            config = ConfigParser.SafeConfigParser()
            success = config.read(littlechef.CONFIGFILE)
            if success:
                if os.path.isfile("rackspace.yaml"):
                    return yaml.load(file("rackspace.yaml"))
                elif os.path.isfile("rackspace.yml"):
                    return yaml.load(file("rackspace.yml"))
                else:
                    print(("WARNING: Reading configuration from deprecated "
                           "{0} file, consider upgrading to use "
                           "rackspace.yaml")
                          .format(littlechef.CONFIGFILE))
                    return dict(config.items('rackspace'))

            else:
                abort("Could not read littlechef configuration file! "
                      "Make sure you are running in a "
                      "kitchen (fix new_kitchen).")
        except ConfigParser.ParsingError:
            pass
        except ConfigParser.NoSectionError:
            pass

        return None

    def __init__(self, options=None):
        self.command_classes = get_command_classes()
        if options is None:
            options = self._read_littlechef_config()

        self.options = options or {}

    def _read_secrets_file(self, secrets_file):
        if secrets_file:
            try:
                del self.options['secrets-file']
                secrets_file = os.path.expanduser(secrets_file)
                # yes, look at another config file
                secrets_config = ConfigParser.SafeConfigParser()
                secrets_config.read(secrets_file)
                # assert _
                return dict(secrets_config.items(ConfigParser.DEFAULTSECT))
            except ConfigParser.ParsingError:
                pass

        return {}

    def get_api(self):
        username = self.options.get('username')
        key = self.options.get('key')
        region = self.options.get('region', '').lower()

        if region not in ['dfw', 'ord', 'syd', 'lon', 'iad', 'hkg']:
            abort(FailureMessages.INVALID_REGION)

        return RackspaceApi(username=username, key=key, region=region)

    def get_deploy(self):
        key_filename = self.options.get("private_key", "~/.ssh/id_rsa")
        return ChefDeployer(key_filename=key_filename)

    def _expand_argument(self, args, key):
        if args.get(key) and not isinstance(args.get(key), list):
            args[key.replace('-', '_')] = args[key].split(',')

    def main(self, cmd_args):
        (options, args) = parser.parse_args(cmd_args)

        if not args:
            raise InvalidCommand

        user_command = args[0]
        templates = args[1:]

        matched_commands = filter(lambda command_class:
                                  command_class.name == user_command,
                                  self.command_classes)

        if not matched_commands:
            raise InvalidCommand

        for k, v in vars(options).items():
            if v is not None and v != '':
                self.options[k] = v

        config_templates = self.options.get('templates', {})
        if 'templates' in self.options:
            del self.options['templates']

        for template in templates:
            if template not in config_templates:
                raise InvalidTemplate

            template_arguments = config_templates.get(template)
            for key, value in template_arguments.iteritems():
                if key not in self.options:
                    self.options[key] = value
                elif isinstance(self.options[key], list):
                    self.options[key] += value
                else:
                    self.options[key] = value

        self.options.update(self._read_secrets_file(
                            self.options.get('secrets-file')))

        command_class = matched_commands[0]
        command_kwargs = {'rackspace_api': None,
                          'chef_deployer': None}

        if command_class.requires_api:
            command_kwargs['rackspace_api'] = self.get_api()
        if command_class.requires_deploy:
            command_kwargs['chef_deployer'] = self.get_deploy()

        command = command_class(**command_kwargs)

        args = self.options

        if not command.validate_args(**args):
            abort(FailureMessages.MISSING_REQUIRED_ARGUMENTS)
            return

        public_key = args.get('public_key', "~/.ssh/id_rsa.pub")
        args['public_key_file'] = file(os.path.expanduser(public_key))

        self._expand_argument(args, 'runlist')
        self._expand_argument(args, 'plugins')
        self._expand_argument(args, 'post-plugins')
        self._expand_argument(args, 'networks')

        if 'use-opscode-chef' in args:
            args['use_opscode_chef'] = bool(args['use-opscode-chef'])

        if user_command == 'create' and 'networks' in args:
            if '00000000-0000-0000-0000-000000000000' not in args['networks']:
                raise InvalidConfiguration(
                    FailureMessages.MUST_SPECIFY_PUBLICNET
                )

        command.execute(**args)


class MissingRequiredArguments(Exception):
    pass


class InvalidConfiguration(Exception):
    pass


class InvalidCommand(Exception):
    pass


class InvalidTemplate(Exception):
    pass
